### ISSUE TITLE
something went wrong in DEVELOPMENT when using $variable in scss

### ISSUE THREAD

**Issue by scratchoo** [rails/rails#38792]
First of all, an example where this scenario can happen:
You are using `'bootstrap', '~> 4.3', '>= 4.3.1' `  in your Gemfile and your intent is to have a "dummy.scss" where you are using some bootstrap variable like for example **$red**

Then in **application.scss** you want to do something like:

```
@import "bootstrap"
@import "dummy" // <----- THIS is where the $red variable is.
```
**importing "bootstrap" and/or "dummy" is optional** 

Here is a demo app to see the error in action: https://github.com/scratchoo/applika

### Steps to reproduce
1) Create a file dummy.scss under assets/stylesheets  (**assets/stylesheets/dummy.scss**)

2) add the following:

`.element{ background: $red}`

3) Rename application.css to application.scss and **delete everything written in that file** (**keep it empty or use the @import syntax**, i.e: `@import "bootstrap";`)

> Note: deleting the content in application.scss is key, because if you keep the (require_tree etc..) syntax:
> ```
>  /*
>  *= require_tree .
>  *= require_self
>  */
> ``` 
> The error will be shown clearly (Error: Undefined variable: "$red".) but for someone who is going to use `@import....` then no such clear error will appear.

4) Run `rails s`

### Expected behavior
It should work l think?!!

### Actual behavior

**We're sorry, but something went wrong.**
**If you are the application owner check the logs for more information.**

Weirder... if you keep refreshing using CTRL+R or the keyboard f5 button, the page above will disappear and another error shows up:

```
**Sprockets::Rails::Helper::AssetNotPrecompiled** in CustomSessions#new
Asset `application.css` was not declared to be precompiled in production.
Declare links to your assets in `app/assets/config/manifest.js`.

  //= link application.css
and restart your server
```
### The solution
After 1 day of crazy tests, I found out that there is a file under assets/config called manifest.js with the following line:

`//= link_directory ../stylesheets .css`

From the errors I described above, I concluded that somehow all css (or scss) files are read/loaded/compiled? separately, and when the variable $red is found in dummy.scss there is no way it knows that variable is declared in another file.

**So I created a folder assets/stylesheets/a_dummy_folder and put that dummy.scss on it, Now it works!** 

I am not sure this is the expected behavior **BUT... the error page I was getting is totally confusing, it made me crazy!**

### System configuration
**Rails version**:
Rails 6.0.2.2
**Ruby version**:
ruby '2.7.0'

**Comment by scratchoo** [comment]
I was playing with the demo app I created (which I linked to above) and it seems to have a weird behavior (maybe a little different) than the original project I was working on.

Watch this fully: https://imgur.com/yKdQdrB

Note that when I cut off the server, I do it in purpose because the "something went wrong" page won't show up until you edit something and you restart the server,  once that page appears, I refresh and the page error disappear, but in my original project that page doesn't disappear even after multiple refreshes!!! I am so confused.

**Comment by eugeneius** [comment]
The "something went wrong" part of this has been fixed in https://github.com/rails/rails/pull/36521, but hasn't been released yet. If you point your Gemfile at the 6-0-stable branch, you'll see the 'Undefined variable: "$red"' error on the first request.

The other issue looks like https://github.com/rails/sprockets/issues/655 - I confirmed that downgrading your demo app to Sprockets 3.x fixes it. I'm going to close this since it doesn't look like there's anything to fix in Rails itself here.
